gh-42765 timeit.Timer Documentation Update

richardhob opened the pull request:
Update the `timeit` documentation based on discussion in issue #42765.          
                                                                                
Resolve Issue #42765                                                            
                                                                                
This issue discusses the documentation for the `timeit.Timer` class, as well as possible improvements. `elibendersky` provided a patch to update the  documentation, and `rurpy` improved on this in a post. This merge request  implements `rurpy`'s version of the `timeit.Timer` class documentation with some light edits.                                                                    
                                                                                
The goal of this documentation change is to better explain how to use `timeit.Timer`, what the input arguments are, and how the args are use to actually time the execution of the provided snippet.

<!-- gh-issue-number: gh-42765 -->
* Issue: gh-42765
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
:books: Documentation preview :books:: https://cpython-previews--108744.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

richardhob committed:
Add relavent news message using blurb

richardhob committed:
Update timeit.Timer documentation.

Documentations updates are based on Issue 42765 discussion, with the
intent of better describing the Timer class, how to use it, limitations
of the parameters, etc.

richardhob committed:
Add blurb for the New Documentation with light details about the  documentation update

richardhob committed:
Fix RST in news blurb - use two back ticks instead of one

encukou committed:
gh-137210: Move PYTHON_API_VERSION & PYTHON_ABI_VERSION to patchlevel.h (GH-138545)

Move PYTHON_API_VERSION & PYTHON_ABI_VERSION to patchlevel.h

This allows removing #include "modsupport.h" from python_ver_rc.h,
which allows modsupport.h to use common helpers from Python.h --
specifically, `_Py_PACK_VERSION` for defining limited API.

picnixz committed:
gh-138633: fix Sphinx references in `ssl` documentation (#138648)

adorilson committed:
Doc: Apply keyword role around finally (in sys module doc) (GH-138544)

yihong0618 committed:
gh-138318, PyREPL: builtins should not be highlighted when used as attribute names (#138319)

wiomoc committed:
gh-57911: Fix failing symlink test in tarfile (GH-138626)

sobolevn committed:
gh-138644: Update c-api docs of `PyInterpreterState` about PEP-684 (#138651)

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

axdeyy committed:
gh-138659: Typo in the gc module docstring (#138660)

docs(gc): fix typo in get_threshold() docstring

Removes a duplicate "the" from the docstring for the `gc.get_threshold()` function.

zklaus committed:
gh-133143: Add sys.abi_info (GH-137476)

This makes information about the interpreter ABI more accessible.

Co-authored-by: Petr Viktorin <encukou@gmail.com>
Co-authored-by: Victor Stinner <vstinner@python.org>
Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

ashm-dev committed:
gh-138665: Move `platform.invalidate_caches` docs (#138667)

bmerry committed:
gh-136234: Fix `SelectorSocketTransport.writelines` to be robust to connection loss (#136743)

picnixz committed:
gh-133879: Improve consistency of full stops and whitespace in What's New (#138635)

AA-Turner committed:
gh-133143: Condense the implementation for ``sys.abi_info`` (#138672)

AA-Turner committed:
gh-101100: Resolve reference warnings in reference/ (#138418)

DinoV committed:
gh-138679: Opcodes which consume no inputs should indicate they produced the val… (#138678)

Opcodes which consume no inputs should indicate they produced the value, not an arbitrary local

AA-Turner committed:
GH-101100: Remove some entries from ``nitpick_ignore`` (#138464)

vstinner committed:
gh-137242: Add a --no-randomize option, and use it in Android CI (#138649)

Adds a --no-randomize option to the CI runner, so that randomisation can be easily
disabled for --fast-ci and --slow-ci configurations on single-threaded testing platforms
like Android, iOS, and Emscripten.

---------

Co-authored-by: Malcolm Smith <smith@chaquo.com>

pablogsal committed:
gh-138385: Sample all interpreters in the tachyon profiler (#138398)

corona10 committed:
gh-137838: Fix JIT trace buffer overrun by increasing possible exit stubs (gh-138177)

jbosboom committed:
gh-138535: Optimize fill_time for typical timestamps (#138537)

While file timestamps can be anything the file system can store, most
lie between the recent past and the near future.  Optimize fill_time()
for typical timestamps in three ways:

- When possible, convert to nanoseconds with C arithmetic.
- When using C arithmetic and the seconds member is not required (for
  st_birthtime), avoid creating a long object.
- When using C arithmetic, reorder the code to avoid the null checks
  implied in Py_XDECREF().

Co-authored-by: Victor Stinner <vstinner@python.org>

skirpichev committed:
Add impl-detail block for PyLong_FromLong docs (#126422)

Co-authored-by: Brian Schubert <brianm.schubert@gmail.com>

yagggi committed:
gh-138577: Mention Unix-specific limitations of `getpass.getpass(echo_char=...)` (#138677)

In bf8bbe9a813dd9fc2dd14be06df172b7d26ca1af, `getpass.getpass` gained
the ability to provide keyboard feedback through `echo_char`.

On Unix, line editing shortcuts such as Ctrl+U were previously handled
as the terminal operates in canonical mode (see termios(3)). However,
since keyboard feedback requires to switch to noncanonical mode, this
now results in an inconsistency when `getpass.getpass` uses `echo_char`
as those shortcuts are no more supported. This limitation is specific
to Unix and does not affect Windows users where line editing shortcuts
were never supported.

encukou committed:
gh-133143: Use _Py_ID for the other literals in sys (GH-138698)

Follow-up refactoring after GH-133143, where we use `_Py_ID` for "big" and "little"
in `abi_info.byteorder`.
This uses `_Py_ID` for `sys.byteorder`, but also `float_repr_style` and a module name.

vstinner committed:
gh-138535: Pass directly state to posix fill_time() (#138693)

Pass directly the module state instead of passing the module to avoid
a call to get_posix_state().

serhiy-storchaka committed:
gh-88886: Remove excessive encoding name normalization (GH-137167)

The codecs lookup function now performs only minimal normalization of
the encoding name before passing it to the search functions:
all ASCII letters are converted to lower case, spaces are replaced
with hyphens.

Excessive normalization broke third-party codecs providers, like
python-iconv.

Revert "bpo-37751: Fix codecs.lookup() normalization (GH-15092)"

This reverts commit 20f59fe1f7748ae899aceee4cb560e5e1f528a1f.

kj7rrv committed:
gh-85076: Document exceptions that can be raised by importlib.import_module (GH-94662)

sobolevn committed:
gh-136616: Improve `assert` syntax error messages (#136653)

lkollar committed:
gh-135953: Add flamegraph reporter to sampling profiler (#138715)

pablogsal committed:
gh-135953: Fix refleak in cache method (#138721)

DinoV committed:
gh-138230: Remove dead code in code gen - codegen_check_annotation is only calle… (#138228)

Remove dead code in code gen - codegen_check_annotation is only called if future annotations are enabled, and if future annotations are enabled it does nothing.

picnixz committed:
gh-137371: refactor and fortify `test.support.hashlib_helper` (#137375)

picnixz committed:
gh-134531: use `EVP_MAC` API for `_hashlib.HMAC` when possible (#135235)

Use `EVP_MAC` API for HMAC on builds with OpenSSL 3.0 or later.

picnixz committed:
gh-136006: fix `Py_NAN` expansion on Solaris systems (#136575)

On Solaris, `Py_NAN` may expand as a function address instead of a floating-point number.
This amends commit 7a3b03509e5e3e72d8c47137579cccb52548a318.

magnified103 committed:
gh-138008: Fix segfaults in _ctypes due to invalid argtypes (GH-138285)


Signed-off-by: Nguyen Viet Dung <29406816+magnified103@users.noreply.github.com>
Signed-off-by: Nguyen Viet Dung <dung@ekluster.com>
Co-authored-by: blurb-it[bot] <43283697+blurb-it[bot]@users.noreply.github.com>
Co-authored-by: Yongzi Li <204532581+Yzi-Li@users.noreply.github.com>

sobolevn committed:
gh-138716: Fix `assert a := b` syntax error message (#138718)

sobolevn committed:
gh-138729: Cover `inspect.formatannotationrelativeto` with tests (#138730)

JelleZijlstra committed:
gh-138349: Fix crash when combining module-level annotation and listcomp (#138363)

hugovk committed:
gh-138744: GitHub Actions: pin to `windows-2022` (#138743)

AA-Turner committed:
GH-138562: Remove ``sort()`` from the common sequence methods in the data model (#138563)

tungol committed:
gh-138432: Improved invalid path checking in zoneinfo.reset_tzpath() (GH-138433)

* Improve error messages for path-like relative paths and path-like bytes paths.
* TZPATH is now always a tuple of strings.

picnixz committed:
gh-116946: add `Py_TPFLAGS_IMMUTABLETYPE` to several internal types (#138582)

The following types are now immutable:

* `_curses_panel.panel`,
* `[posix,nt].ScandirIterator`, `[posix,nt].DirEntry` (exposed in `os.py`),
* `_remote_debugging.RemoteUnwinder`,
* `_tkinter.Tcl_Obj`, `_tkinter.tkapp`, `_tkinter.tktimertoken`,
* `zlib.Compress`, and `zlib.Decompress`.

yihong0618 committed:
gh-138081: fix some dead links in InternalDocs (#138082)

skirpichev committed:
gh-71810: Fix corner case (length==0) for int.to_bytes() (#138739)

```pycon
>>> (0).to_bytes(0, 'big', signed=True)
b''
>>> (-1).to_bytes(0, 'big', signed=True)  # was b''
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    (-1).to_bytes(0, 'big', signed=True)
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
OverflowError: int too big to convert
```

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

ZeroIntensity committed:
gh-138479: Ensure that `__typing_subst__` returns a tuple (GH-138482)

Raise an exception if __typing_subst__ returns a non-tuple object.

---------

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

cmaloney committed:
gh-138013: Move I/O test infrastructre to test_io.utils (#138475)

This moves test support code to `test_io.utils` via copy/paste
code movement then adjusts imports as needed
(remove unneded + add all required).

vstinner committed:
gh-138756: Fix memory leak in PyInitConfig_Free() (#138759)

Clear also memory of PyConfig members.

picnixz committed:
gh-116946: fully implement GC protocol for `_tkinter.Tk{app,tt}Object` (#138331)

pablogsal committed:
gh-138704: Add PermissionError messages to profiling.sampling (#138705)

yoney committed:
gh-116738: add multi-threaded tests for `resource` module on free-threading builds (#138504)

maurycy committed:
gh-138005: Document that CSV `skipinitialspace=True` and `delimiter=' '` require quotation for empty fields (#138006)

benjaminp committed:
closes gh-138706: update Unicode to 17.0.0 (#138719)

ZeroIntensity committed:
Revert "gh-116946: fully implement GC protocol for `_tkinter.Tk{app,tt}Object` (#138331)" (#138807)

This reverts commit 283380a9417eebf1bf9d3a1e941cef94149712bb.

wiomoc committed:
Remove unused prototype `_PyPegen_set_source_in_metadata` (#138810)

Remove unused prototype

chris-eibl committed:
gh-138801: re-enable `pyrepl` tests on Windows (#138802)

JanEricNitschke committed:
gh-138669: Increase test coverage for difflib (GH-138670)

jxes993409 committed:
gh-138163: skip failures if tests are run with `SCHED_BATCH` on glibc (GH-138576)

vstinner committed:
gh-138756: Fix leak of inittab memory in PyInitConfig_Free() (GH-138792)

vstinner committed:
gh-129813, PEP 782: Add PyBytesWriter C API (#138822)

ivonastojanovic committed:
Improve permission error messages in pdb and asyncio.tools (#134290)

vstinner committed:
gh-129813, PEP 782: Add PyBytesWriter_Format() (#138824)

Modify PyBytes_FromFormatV() to use the public PyBytesWriter API
rather than the _PyBytesWriter private API.

kumaraditya303 committed:
gh-138661: fix data race in `PyCode_Addr2Line` (#138664)

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _hashopenssl (#138835)

Replace PyBytes_FromStringAndSize(NULL, size) with the new public
PyBytesWriter API.

StanFromIreland committed:
gh-71648: Remove unused image `Doc/faq/python-video-icon.png` (GH-138841)

wiomoc committed:
gh-138804: Check type in `shlex.quote` (GH-138809)

StanFromIreland committed:
gh-71648: Optimize PNG files (GH-138842)

* Only keep > 1 kb savings

barneygale committed:
GH-128520: pathlib ABCs: improve protocol for 'openable' objects (#134101)

Rename `pathlib._os.magic_open()` to `vfsopen()`. The new name is a bit
less abstract, and it aligns with the `vfspath()` method added in 5dbd27d.

Per discussion on discourse[^1], adjust `vfsopen()` so that the following
methods may be called:

- `__open_reader__()`
- `__open_writer__(mode)`
- `__open_updater__(mode)`

These three methods return readable, writable, and full duplex file objects
respectively. In the 'writer' method, *mode* is either 'a', 'w' or 'x'. In
the 'updater' method, *mode* is either 'r' or 'w'.

In the pathlib ABCs, replace `ReadablePath.__open_rb__()` with
`__open_reader__()`, and replace `WritablePath.__open_wb__()` with
`__open_writer__()`.

[^1]: https://discuss.python.org/t/open-able-objects/90238

Co-authored-by: Petr Viktorin <encukou@gmail.com>

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in binascii (#138825)

Replace the private _PyBytesWriter API with the new public
PyBytesWriter API.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in pickle and struct (#138833)

Replace the private _PyBytesWriter API with the new public
PyBytesWriter API.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _PyBytes_DecodeEscape2() (#138838)

Replace the private _PyBytesWriter API with the new public
PyBytesWriter API.

vstinner committed:
gh-129813, PEP 782: Use Py_GetConstant(Py_CONSTANT_EMPTY_BYTES) (#138830)

Replace PyBytes_FromStringAndSize(NULL, 0) with
Py_GetConstant(Py_CONSTANT_EMPTY_BYTES). Py_GetConstant() cannot
fail.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in socket and mmap (#138831)

Replace PyBytes_FromStringAndSize(NULL, size) and _PyBytes_Resize()
with the new public PyBytesWriter API.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in memoryview (#138836)

Replace PyBytes_FromStringAndSize(NULL, size) with the new public
PyBytesWriter API.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in posix extension (#138829)

Replace PyBytes_FromStringAndSize(NULL, size) and _PyBytes_Resize()
with the new public PyBytesWriter API.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _PyBytes_FromList() (#138837)

Use the new public PyBytesWriter API in:

* _PyBytes_FromHex()
* _PyBytes_FromBuffer()
* _PyBytes_FromList()
* _PyBytes_FromTuple()
* _PyBytes_FromIterator()

Add _PyBytesWriter_ResizeAndUpdatePointer() and
_PyBytesWriter_GetAllocated() helper functions.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in lzma and zlib (#138832)

Replace PyBytes_FromStringAndSize(NULL, size) with the new public
PyBytesWriter API.

vstinner committed:
gh-71810: Fix _PyLong_AsByteArray() undefined behavior (#138873)

Don't read p[-1] when p is an empty string: when n==0.

sobolevn committed:
gh-136355: Deprecate `-b` and `-bb` CLI flags in 3.15 (#136363)

Co-authored-by: Victor Stinner <vstinner@python.org>
Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

pablogsal committed:
gh-135953: Prevent overflow in the profiler flamegraph hover boxes (#138880)

lkollar committed:
gh-135953: Reduce memory usage of stack collectors (#138875)

The stack collector base class keeps all frames until export() is
called, which causes significant unnecessary memory usage. Instead, we
can process the frames on the fly in the collect call by dispatching the
aggregation logic to the subclass through the process_frames method.

Co-authored-by: Pablo Galindo Salgado <pablogsal@gmail.com>

kianelbo committed:
gh-138871: Clarify NameError exception in 'del' (#138881)

johnslavik committed:
Fix typo in stack_protection.md (GH-138876)

sobolevn committed:
gh-138886: Remove deprecated `PySys_ResetWarnOptions` C-API function (#138887)

pablogsal committed:
gh-138794: Communicate to PyRefTracer when they are being replaced (#138797)

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _PyBytes_FormatEx() (#138839)

Replace the private _PyBytesWriter API with the new public
PyBytesWriter API.

pablogsal committed:
gh-137992: Stop the world when calling PyRefTracer_SetTracer (#137994)

F0RRZZ committed:
gh-138902: Fix generator send arg name (#138905)

Co-authored-by: marat <khagazheev@artlebedev.ru>

abdoulrasheed committed:
gh-137017: Ensure `Thread.is_alive()` only returns False after the underlying OS thread exits (gh-137315)

encukou committed:
gh-137956: Guard against non-free-threaded extensions in free-threaded builds (GH-137957)

serhiy-storchaka committed:
gh-138779: Use the dev_t converter for st_rdev (GH-138780)

This allows to support device numbers larger than 2**63-1.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in fcntl (#138921)

Replace PyBytes_FromStringAndSize(NULL, size) with the new public
PyBytesWriter API.

Don't build the fcntl with the limited C API anymore, since
the PyBytesWriter API is not part of the limited C API.

vstinner committed:
gh-129813, PEP 782: Init small_buffer in PyBytesWriter_Create() (#138924)

Fill small_buffer with 0xFF byte pattern to detect the usage of
uninitialized bytes in debug build.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _curses (#138920)

Replace PyBytes_FromStringAndSize(NULL, size) and _PyBytes_Resize()
with the new public PyBytesWriter API.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _codecs.escape_decode() (#138919)

Replace PyBytes_FromStringAndSize(NULL, size) and _PyBytes_Resize()
with the new public PyBytesWriter API.

StanFromIreland committed:
gh-134953: Expand theming for `True`/`False`/`None` (#135000)

Co-authored-by: Łukasz Langa <lukasz@langa.pl>

serhiy-storchaka committed:
gh-138712: Add os.NODEV (GH-138728)

kumaraditya303 committed:
gh-137992: fix `PyRefTracer_SetTracer` to start world before returning  (#138925)

fix deadlock in PyRefTracer_SetTracer

savannahostrowski committed:
GH-132732: Use pure op machinery to optimize various instructions with `_POP_TOP` and `_POP_TWO`  (#137577)

savannahostrowski committed:
GH-132732: Remove textwrap import (#138933)

serhiy-storchaka committed:
gh-137490: Fix signal.sigwaitinfo() on NetBSD (GH-137523)

Handle ECANCELED in the same way as EINTR to work around the Posix
violation in the NetBSD's implementation.

00ll00 committed:
gh-138239: Fix incorrect highlighting of "type" in type statements in the REPL (GH-138241)

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

Fidget-Spinner committed:
gh-132732: Clear errors in JIT optimizer on error (GH-136048)

yihong0618 committed:
gh-128636: Fix crash in PyREPL when os.environ is overwritten with an invalid value for macOS (GH-138089)

Signed-off-by: yihong0618 <zouzou0208@gmail.com>
Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

ambv committed:
gh-134953: Make the True/False/None check more efficient (GH-138931)

serhiy-storchaka committed:
gh-37817: Allow assignment to __bases__ of direct subclasses of builtin classes (GH-137585)

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _hashopenssl (#138922)

Replace PyBytes_FromStringAndSize(NULL, size) with the new public
PyBytesWriter API.

loic-simon committed:
gh-69605: Hardcode some stdlib submodules in PyREPL module completion (os.path, collections.abc...) (GH-138268)

Co-authored-by: Łukasz Langa <lukasz@langa.pl>

mehdihassan93 committed:
gh-97517: Add documentation links to datetime strftime/strptime docstrings (#138559)

* Add documentation links to datetime strftime/strptime docstrings

- Add links to format codes documentation for all strftime methods
- Add links to format codes documentation for all strptime methods
- Addresses issue #97517

* Update C extension docstrings with format codes documentation

* Regenerate clinic code for updated docstrings

* Add clinic-generated header file for updated docstrings

* Fix docstring spacing consistency in both Python and C files

* Update Lib/_pydatetime.py

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

---------

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _winapi.PeekNamedPipe() (#138930)

Replace PyBytes_FromStringAndSize(NULL, size) and _PyBytes_Resize()
with the new public PyBytesWriter API.

cmaloney committed:
gh-129813, PEP 782: Use PyBytesWriter in _io.FileIO.readall (#138901)

Performance about the same, using the benchmark from gh-120754:

```
pyperf compare_to main.json pep782.json
read_file_small: Mean +- std dev: [main] 5.71 us +- 0.05 us -> [pep782] 5.68 us +- 0.05 us: 1.01x faster
read_file_large: Mean +- std dev: [main] 89.7 us +- 0.9 us -> [pep782] 86.9 us +- 0.8 us: 1.03x faster
read_all_rst_bytes: Mean +- std dev: [main] 926 us +- 8 us -> [pep782] 920 us +- 12 us: 1.01x faster
read_all_rst_text: Mean +- std dev: [main] 2.24 ms +- 0.02 ms -> [pep782] 2.17 ms +- 0.04 ms: 1.03x faster

Benchmark hidden because not significant (1): read_all_py_bytes

Geometric mean: 1.01x faster
```

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _sha3 (#138923)

Replace PyBytes_FromStringAndSize(NULL, size) with the new public
PyBytesWriter API.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in FileIO.read() (#138955)

Replace PyBytes_FromStringAndSize(NULL, size) and _PyBytes_Resize()
with the new public PyBytesWriter API.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in bufferedio.c (#138954)

Replace PyBytes_FromStringAndSize(NULL, size) and _PyBytes_Resize()
with the new public PyBytesWriter API.

vstinner committed:
gh-129813, PEP 782: Add PyBytesWriter.overallocate (#138941)

Disable overallocation in _PyBytes_FormatEx() at the last write.

emmatyping committed:
Re-order Modules/makesetup in CODEOWNERS and update owners (#138628)

* Re-order Modules/makesetup in CODEOWNERS and add Adam Turner
* Add Erlend to Modules/makesetup in CODEOWNERS

yihong0618 committed:
gh-138958: remove unused variable in Lib/asyncio/streams.py (#138959)

colesbury committed:
gh-137433: Fix deadlock with stop-the-world and daemon threads (gh-137735)

There was a deadlock originally seen by Memray when a daemon thread
enabled or disabled profiling while the interpreter was shutting down.
I think this could also happen with garbage collection, but I haven't
seen that in practice.

The daemon thread could be hung while trying acquire the global rwmutex
that prevents overlapping global and per-interpreter stop-the-world events.
Since it already held the main interpreter's stop-the-world lock, it
also deadlocked the main thread, which is trying to perform interpreter
finalization.

Swap the order of lock acquisition to prevent this deadlock.
Additionally, refactor `_PyParkingLot_Park` so that the global buckets
hashtable is left in a clean state if the thread is hung in
`PyEval_AcquireThread`.

cmaloney committed:
gh-138013: Remove load_tests in test_io.test_general (#138771)

Rely on default test discovery.

Validation:
```bash
 # Run before commit
./python -m test test_io -uall,walltime,largefile,cpu,extralargefile -M25G -o --fail-env-changed -j0 --list-cases | sort > old_cases.txt
 # Run after commit
./python -m test test_io -uall,walltime,largefile,cpu,extralargefile -M25G -o --fail-env-changed -j0 --list-cases | sort > new_cases.txt

diff new_cases.txt old_cases.
 # <outputs no changes in case list>
```

vstinner committed:
gh-129813, PEP 782: Soft deprecate _PyBytes_Resize() (#138964)

Soft deprecate _PyBytes_Resize() and
PyBytes_FromStringAndSize(NULL, size).

StanFromIreland committed:
gh-88473: Implement fast path in date.today() for date types (#130980)

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

brandtbucher committed:
GH-129805: Clean up some changes from GH-129806 (GH-133540)

yihong0618 committed:
gh-135329: prevent infinite traceback loop on Ctrl-C  for strace (GH-138133)

Signed-off-by: yihong0618 <zouzou0208@gmail.com>
Co-authored-by: dura0ok <slpmcf@gmail.com>
Co-authored-by: graymon <greyschwinger@gmail.com>
Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>
Co-authored-by: Łukasz Langa <lukasz@langa.pl>

ivanbelenky committed:
Synced docs and docstring  for `sysconfig.get_platform` (#135530)

Co-authored-by: Mikhail Efimov <efimov.mikhail@gmail.com>
Co-authored-by: Ned Deily <nad@python.org>

iritkatriel committed:
gh-135629: rewrite language reference section on except* to improve clarity (#136150)

aisk committed:
fix some typos (#138977)

zooba committed:
Adds support for building with VisualStudioVersion = 18.0 (GH-138980)

vstinner committed:
gh-129813, PEP 782: Add doc reference link (#138986)

bkjoh committed:
gh-138514: getpass: restrict `echo_char` to a single ASCII character (#138591)

This amends commit bf8bbe9a813dd9fc2dd14be06df172b7d26ca1af by
restricting `echo_char` in `getpass.getpass` to single printable
ASCII characters as it would be uncommon to use long strings or
multi-byte characters for keyboard feedback.

---------

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>
Co-authored-by: Brian Schubert <brianm.schubert@gmail.com>

JelleZijlstra committed:
gh-137226: Fix get_type_hints() on generic TypedDict with stringified annotations (#138953)

This issue appears specifically for TypedDicts because the TypedDict constructor
code converts string annotations to ForwardRef objects, and those are not evaluated
properly by the get_type_hints() stack because of other shenanigans with type
parameters.

This issue does not affect normal generic classes because their annotations are not
pre-converted to ForwardRefs.

The fix attempts to restore the pre- #137227 behavior in the narrow scenario where
the issue manifests. It mostly makes changes only in the paths accessible from get_type_hints(),
ensuring that newer APIs (such as evaluate_forward_ref() and annotationlib) are not affected
by get_type_hints()'s past odd choices. This PR does not fix issue #138949, an older issue I
discovered while playing around with this one; we'll need a separate and perhaps more
invasive fix for that, but it should wait until after 3.14.0.

AlexWaygood committed:
Revert "gh-118803: Remove `ByteString` from `typing` and `collections.abc` (#118804)" (#138990)

serhiy-storchaka committed:
gh-134716: Support regular expressions in -W and PYTHONWARNINGS (GH-138149)

serhiy-storchaka committed:
gh-71679: Share the repr implementation between bytes and bytearray (GH-138181)

This allows to use the smart quotes algorithm in the bytearray's repr.

codenamenam committed:
gh-138966: Add description for `gi_suspended` attribute (GH-139008)

AA-Turner committed:
GH-123299: Copyedit 3.14 What's New: C API (#138987)

ryanking13 committed:
gh-99948: Support ctypes.util.find_library in emscripten environment (#138519)

Adds support for ctypes loading .so files directly.

---------

Co-authored-by: blurb-it[bot] <43283697+blurb-it[bot]@users.noreply.github.com>
Co-authored-by: Russell Keith-Magee <russell@keith-magee.com>

StanFromIreland committed:
gh-138998: Upgrade vendored expat to 2.7.2 (#138999)

ambv committed:
gh-134466: Don't run when termios is inaccessible (GH-138911)

Without the ability to set required capabilities, the REPL cannot
function properly (syntax highlighting and multiline editing can't
work).

We refuse to work in this degraded state.

picnixz committed:
gh-76760: test that `uuid.uuid1()` sets the version field (#139033)

adang1345 committed:
gh-138896: Fix error installing C runtime on non-updated Windows machines (GH-138932)

encukou committed:
gh-107862: Add property-based round-trip tests for base64 (#119406)

* Add property-based tests to test_base64

* Allow multiple positional arguments in @hypothesis.example stub

* Simplify the altchars strategy

PrinceNaroliya committed:
gh-139006: Doc: Clarify html.escape function description (#139016)

Doc: Clarify html.escape function description (GH-139006)

hoodmane committed:
gh-128627: Use __builtin_wasm_test_function_pointer_signature for Emscripten trampoline (#137470)

With https://github.com/llvm/llvm-project/pull/150201 being merged, there is 
now a better way to generate the Emscripten trampoline, instead of including 
hand-generated binary WASM content. Requires Emscripten 4.0.12.

iamrajhans committed:
gh-129368: In PyRun C API docs, clarify what a "start token" is (#129935)

Aniketsy committed:
gh-138952:  Document platform.machine() output casing inconsistency across platforms (#138962)

Co-authored-by: AN Long <aisk@users.noreply.github.com>

DinoV committed:
gh-131776: Expose functions called from the interpreter loop via PyAPI_FUNC (#134242)

ZeroIntensity committed:
gh-128639: Don't assume one thread in subinterpreter finalization with fixed daemon thread support (GH-134606)

This reapplies GH-128640.

PrinceNaroliya committed:
gh-137988: Fix const description in argparse.add_argument() docs (#138315)

Co-authored-by: Savannah Bailey <savannahostrowski@gmail.com>

pganssle committed:
Add pganssle to CODEOWNERS for ``zoneinfo`` (#139032)

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

AA-Turner committed:
GH-92266: Remove embedded tabs from ``c-analyzer/cpython/_parser.py`` (#137622)

hansthen committed:
gh-132558: Improve `argparse` docs on combining `type` and `choices` (#133827)

Co-authored-by: Savannah Bailey <savannahostrowski@gmail.com>

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _Py_bytes_maketrans() (#139044)

Replace PyBytes_FromStringAndSize(NULL, size) with the new public
PyBytesWriter API.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _multiprocessing (#139047)

Replace PyBytes_FromStringAndSize(NULL, size) and _PyBytes_Resize()
with the new public PyBytesWriter API.

Change also 'read' variable type from int to Py_ssize_t.

chirizxc committed:
gh-138736: Fix sqlite3.Connection.blobopen 'row' parameter type and naming (#138738)

The 'row' parameter represents a ROWID. Clarify this in docs and docstrings.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _testclinic (#139048)

Replace PyBytes_FromStringAndSize(NULL, size) with the new public
PyBytesWriter API.

vstinner committed:
gh-129813, PEP 782: Set invalid bytes in PyBytesWriter (#139054)

Initialize the buffer with 0xFF byte pattern when creating a writer
object, but also when resizing/growing the writer.

moshekaplan committed:
GH-139067: Add example for `argparse`'s `append` action (#131389)

AA-Turner committed:
GH-137630: Convert ``_interpqueues`` to use Argument Clinic (#137685)

DinoV committed:
gh-138714: Don't assume next block has instructions when propagating line numbers (#138770)

Co-authored-by: Irit Katriel <iritkatriel@yahoo.com>

rumpelsepp committed:
gh-86819: Add ISO-TP CAN socket constants (#23794)

Co-authored-by: Stefan Tatschner <stefan@rumpelsepp.org>
Co-authored-by: Patrick Menschel <menschel.p@posteo.de>

corona10 committed:
gh-137838: Move _PyUOpInstruction buffer to PyInterpreterState (gh-138918)

sergey-miryanov committed:
GH-138355: Remove trash_delete_later from _gc_runtime_state (#138767)

Remove trash_delete_later and trash_delete_nesting from _gc_runtime_state.

zeff-ir committed:
gh-138813: Fix mutable default kwargs={} in multiprocessing BaseProcess and DummyProcess to use None (GH-138814)

* gh-138813: Default `BaseProcess` `kwargs` to `None` (#138814)

Set `BaseProcess.__init__(..., kwargs=None)` and initialize `kwargs` with
`dict(kwargs) if kwargs else {}`. This avoids a shared mutable default and
matches threading.Thread behavior.

Co-authored-by: Dmitrii Chuprov <cheese@altlinux.org>

* DummyProcess kwargs=None (which threading.Thread accepts properly)

Co-authored-by: Gregory P. Smith <greg@krypto.org>

rhettinger committed:
gh-139074: Fix missing high precision case in sumprod() (gh-139075)

pablogsal committed:
gh-138970: Add general metadata system to the peg generator (#138971)

olekskrav committed:
Remove unnecessary slice in heapq.py (gh-139087)

rhettinger committed:
Revert gh-139075.  The sumprod fix needs more work. (gh-139092)

rffontenelle committed:
Remove unmatched closing parenthesis (GH-139082)

markshannon committed:
GH-138378: Move globals-to-consts pass into main optimizer pass (GH-138379)

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _socket (#139097)

Replace PyBytes_FromStringAndSize(NULL, size) and _PyBytes_Resize()
with the new public PyBytesWriter API.

erlend-aasland committed:
gh-116943: Ensure makesetup does not choke on missing EOL in Setup.* files (#139100)

vstinner committed:
gh-129813, PEP 782: Optimize byteswriter_resize() (#139101)

There is no need to copy the small_buffer in PyBytesWriter_Create().

cmaloney committed:
gh-138720: Make Buffered closed check match flush (GH-138724)

In `_io__Buffered_flush_impl` the macro `CHECK_CLOSED` is used to check
the `buffered*` is in a good state to be flushed. That differs slightly
from `buffered_closed`.

In some cases, that difference would result in `close()` thinking the
file needed to be flushed and closed while `flush()` thought the file
was already closed.

This could happen during GC and would result in an unraisable exception.

mhsmith committed:
Make Android streams respect the unbuffered (`-u`) option (#138806)

Android pipes stdout/stderr to the log, which means every write to the log
becomes a separate log line. As a result, most practical uses of stdout/stderr
should be buffered; but it doesn't hurt to preserve unbuffered handling in case
it's useful.

corona10 committed:
gh-137838: Add CI for no-opt JIT (#139081)

* gh-137838: Add CI for no-opt JIT

* nit

StanFromIreland committed:
gh-138998: `expat/refresh.sh`: Fail if Step 3 is not completed (GH-139049)

ZeroIntensity committed:
gh-139098: Use multiphase initialization in `_testcapi` (GH-139102)

Use multiphase initialization in the _testcapi module to allow loading in subinterpreters. The isolation here isn't perfect as there's still some use of globals, but _testcapi should generally work in other interpreters.

encukou committed:
gh-138970: Adjust tests for pegen rule flags (#139107)

ZeroIntensity committed:
gh-135729: Store reference to globals in `Interpreter._decref` (GH-139104)

ZeroIntensity committed:
Document `Py_AddPendingCall()` change with subinterpreters in 3.12 (GH-139117)

Prior to 3.9, Py_AddPendingCall() would always run pending calls in the main interpreter, but then each interpreter got their own ceval state, and they were scheduled for any interpreter. In GH-104813, this was undone, so Py_AddPendingCall() would always schedule for the main interpreter.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in ssl.MemoryBIO (#139113)

Replace PyBytes_FromStringAndSize(NULL, size) and _PyBytes_Resize()
with the new public PyBytesWriter API.

ZeroIntensity committed:
gh-136003: Execute pre-finalization callbacks in a loop (GH-136004)

vstinner committed:
gh-135755: Make Py_TAIL_CALL_INTERP macro private (#138981)

Rename Py_TAIL_CALL_INTERP to _Py_TAIL_CALL_INTERP.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _sqlite (#138956)

Replace PyBytes_FromStringAndSize(NULL, size) with the new public
PyBytesWriter API.

Fidget-Spinner committed:
gh-139109: Dynamic opcode targets (GH-139111)

Make opcode targets table dynamic

mhsmith committed:
gh-137242: Mention Android binary releases in documentation (#138305)

Adds a mention of binary releases to the Android documentation.

---------
Co-authored-by: Russell Keith-Magee <russell@keith-magee.com>
Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

AA-Turner committed:
gh-64490: Argument Clinic: Add support for ``**kwds`` (#138344)

This adds a scaffold of support, initially only working with
strictly positional-only arguments. The FASTCALL calling
convention is not yet supported.

skirpichev committed:
gh-73487: Convert `_decimal` to use Argument Clinic (part 7) (#138221)

Use "defining class" converter, where possible.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in _ssl (#138929)

Replace PyBytes_FromStringAndSize(NULL, size) and _PyBytes_Resize()
with the new public PyBytesWriter API.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in bufferedio.c (#139121)

Replace PyBytes_FromStringAndSize(NULL, size) and _PyBytes_Resize()
with the new public PyBytesWriter API.

benjaminp committed:
fixes gh-139090: add os.RWF_DONTCACHE (#139091)

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in socket recvmsg() (#139131)

Replace PyBytes_FromStringAndSize(NULL, size) and _PyBytes_Resize()
with the new public PyBytesWriter API.

ntindle committed:
gh-137976: Explicitly exclude `localtime` from `available_timezones` (#138012)

* fix: available_timezones is reporting an invalid IANA zone name

* 📜🤖 Added by blurb_it.

* correct rst format for backticks

---------

Co-authored-by: blurb-it[bot] <43283697+blurb-it[bot]@users.noreply.github.com>
Co-authored-by: Paul Ganssle <1377457+pganssle@users.noreply.github.com>

AlexWaygood committed:
gh-118803: Improve documentation around `ByteString` deprecation (#139115)

AlexWaygood committed:
gh-118803: Make `ByteString` deprecations louder; remove `ByteString` from `typing.__all__` and `collections.abc.__all__` (#139127)

ZeroIntensity committed:
gh-136003: Skip non-daemon threads when exceptions occur during finalization (GH-139129)

During finalization, we need to mark all non-daemon threads as daemon to quickly shut down threads when sending CTRL^C to the process. This was a minor regression from GH-136004.

da-woods committed:
gh-119494: Fix error messages for deleting/setting type attributes (#119495)

donbarbos committed:
gh-121237: Add `%:z` directive to datetime.strptime (#136961)

vstinner committed:
gh-129813, PEP 782: Fix refleak in sock_recvfrom() (GH-139151)

savannahostrowski committed:
GH-137218: Update `make` for JIT stencils (#137265)

GPHemsley committed:
gh-81148: Eliminate unnecessary check in _strptime when determining AM/PM (#13428)

* bpo-36967: Eliminate unnecessary check in _strptime when determining AM/PM

* Pauls suggestion to refactor test

* Fix test

---------

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>
Co-authored-by: Paul Ganssle <1377457+pganssle@users.noreply.github.com>

ZeroIntensity committed:
gh-126016: Remove bad assertion in `PyThreadState_Clear` (GH-139158)

In the _interpreters module, we use PyEval_EvalCode() to run Python code in another interpreter. However, when the process receives a KeyboardInterrupt, PyEval_EvalCode() will jump straight to finalization rather than returning. This prevents us from cleaning up and marking the thread as "not running main", which triggers an assertion in PyThreadState_Clear() on debug builds. Since everything else works as intended, remove that assertion.

freakboy3742 committed:
gh-138171: Migrate iOS testbed location and add Apple build script (#138176)

Adds tooling to generate and test an iOS XCframework, in a way that will also facilitate
adding other XCframework targets for other Apple platforms (tvOS, watchOS, visionOS and 
even macOS, potentially).

---------
Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

befeleme committed:
Fix the reference to unicode specification (#139138)

serhiy-storchaka committed:
gh-139076: Fix regression in pydoc not showing extension functions (GH-139077)

Fix a bug in the pydoc module that was hiding functions in a Python
module if they were implemented in an extension module and the module did
not have __all__.

lisroach committed:
gh-138310: Adds sys.audit event for import_module (#138311)

* Updates sys.audit calls for imports to include import_module
* Adds unit tests for existing and new functionality

ZeroIntensity committed:
gh-112729: Correctly fail when the process is out of memory during interpreter creation (GH-139164)

pablogsal committed:
gh-138709: Implement CPU time profiling in profiling.sample (#138710)

cmaloney committed:
gh-138013: Split SignalsTest from test_io.test_general (#139079)

Increase parallelism by splitting out `SignalsTest` from test_general.
`SignalsTest` takes 24.2 seconds on my dev machine when fully enabled
making it the largest part of `test_io`. Code move done via copy/paste
then tweak imports.

After splitting `test_io.test_general` is down to 10.1 seconds on my dev
box with all parts enabled.

serhiy-storchaka committed:
Revert "gh-66234: Add flag to disable the use of mmap in dbm.gnu (GH-135005)" (GH-136989)

This reverts commit 0cec424af5904b3d23ad6e3c6d1a27f89d238d64.

StanFromIreland committed:
Link to Discourse in ``.github/CONTRIBUTING.rst`` (#138996)

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

sobolevn committed:
Make sure the ``:keyword:`` role works for ``case`` (#138878)

aisk committed:
gh-138092: Correct the documented signature of ``mmap.flush`` (#138671)

sobolevn committed:
Docs: add missing tools to `Tools/README` (#139150)

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>
Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

pablogsal committed:
gh-138709: Fix race condition in test_external_inspection (#139209)

Fix race condition in test_external_inspection thread status tests

The tests test_thread_status_detection and test_thread_status_gil_detection
had a race condition where the test could sample thread status between when
the sleeper thread sends its "ready" message and when it actually calls
time.sleep(). This caused intermittent test failures where the sleeper
thread would show as running (status=0) instead of idle (status=1 or 2).

The fix moves the thread status collection inside the retry loop and
specifically waits for the expected thread states before proceeding with
assertions. The retry loop now continues until:
- The sleeper thread shows as idle (status=1 for CPU mode, status=2 for GIL mode)
- The busy thread shows as running (status=0)
- Both thread IDs are found in the status collection

This ensures the test waits for threads to settle into their expected states
before making assertions, eliminating the race condition.

pablogsal committed:
gh-138709: Supress stdout/stderr during test_sampling_profiler tests (#139212)

StanFromIreland committed:
`Doc/library/os.rst`: Remove spurious parenthesis (GH-139205)

jaraco committed:
gh-138313: Suppress ResourceWarnings and restore skipped test (#138597)

encukou committed:
gh-136003: Close file descriptors in test (GH-139225)

This fixes file descriptor leaks introduced in GH-136004

StanFromIreland committed:
`Doc/library/zoneinfo.rst`: Fix typo (#139190)

Removes duplicated wording.

savannahostrowski committed:
GH-137218: Fix Makefile to handle FreeBSD (fix for JIT stencil changes) (GH-139170)

KatieLG committed:
gh-95953: Add a css class to changed lines of difflib.HtmlDiff make_table (#139226)

freakboy3742 committed:
gh-136744: Remove unnecessary chmod from pydoc.apropos() test. (GH-136746)

Remove unnecessary chmod from pydoc.apropos() test.

savannahostrowski committed:
GH-139040: Add Dev Container instructions in JIT README (#139041)

vstinner committed:
gh-139156: Use PyBytesWriter in UTF-32 encoder (#139157)

Replace PyBytes_FromStringAndSize() and _PyBytes_Resize() with the
PyBytesWriter API.

vstinner committed:
gh-139156: Use PyBytesWriter in the UTF-7 encoder (#139248)

Replace PyBytes_FromStringAndSize() and _PyBytes_Resize() with the
PyBytesWriter API.

vstinner committed:
gh-139156: Use PyBytesWriter in PyUnicode_AsUnicodeEscapeString() (#139249)

Replace PyBytes_FromStringAndSize() and _PyBytes_Resize() with the
PyBytesWriter API.

vstinner committed:
gh-139156: Use PyBytesWriter in UTF-16 encoder (#139233)

Replace PyBytes_FromStringAndSize() and _PyBytes_Resize() with the
PyBytesWriter API.

vstinner committed:
gh-139156: Use PyBytesWriter in PyUnicode_AsRawUnicodeEscapeString() (#139250)

Replace PyBytes_FromStringAndSize() and _PyBytes_Resize() with the
PyBytesWriter API.

vstinner committed:
gh-129813, PEP 782: Use PyBytesWriter in utf8_encoder() (#138874)

Replace the private _PyBytesWriter API with the new public
PyBytesWriter API in utf8_encoder() and unicode_encode_ucs1().

markshannon committed:
GH-137573: Add test to check that the margin used for overflow protection is larger than the stack space used by the interpreter (GH-137724)

vstinner committed:
gh-139208: Fix regrtest --fast-ci --verbose (#139240)

Don't ignore the --verbose option anymore.

ZeroIntensity committed:
gh-126016: Fix flaky test by allowing the SIGINT return code (GH-139219)

serhiy-storchaka committed:
gh-130567: Fix possible crash in locale.strxfrm() (GH-138940)

On some macOS versions there was an off-by-one error in wcsxfrm() which
caused writing past the end of the array if its size was not calculated
by running wcsxfrm() first.

Co-authored-by: Ronald Oussoren <ronaldoussoren@mac.com>

vstinner committed:
gh-129813, PEP 782: Remove the private _PyBytesWriter API (#139264)

It is now replaced with the new public PyBytesWriter API (PEP 782).

freakboy3742 committed:
gh-136744: Remove a redundant test skip (#139267)

Remove a redundant test skip.

vstinner committed:
gh-74857, PEP 538: Coerce POSIX locale to UTF-8 based locale (#139238)

StanFromIreland commented on Misc/NEWS.d/next/Library/2023-08-21-18-36-09.gh-issue-52769.PoTaac.rst:
This is unrelated?

StanFromIreland commented on Doc/library/timeit.rst:
```diff
@@ -98,47 +98,55 @@ The module defines three convenience functions and a public class:
       .. note::
 
          By default, :meth:`.timeit` temporarily turns off :term:`garbage
-         collection` during the timing.  The advantage of this approach is that
```
Double spaces are fine, please don't do this.

StanFromIreland reviewed (changes requested):
Please remove unrelated changes and see Erlend's comment.

davepeck committed:
gh-138558: Improve handling of Template annotations in annotationlib (#139072)

serhiy-storchaka committed:
gh-67795: Accept any real numbers as timestamp and timeout (GH-139224)

Functions that take timestamp or timeout arguments now accept any
real numbers (such as Decimal and Fraction), not only integers or floats,
although this does not improve precision.

barneygale committed:
GH-139174: Prepare `pathlib.Path.info` for new methods (#139175)

Merge `_WindowsPathInfo` and `_PosixPathInfo` classes into a new
`_StatResultInfo` class. On Windows, this means relying on `os.stat()`
rather than `os.path.isfile()` and friends, which is a little slower. But
there's value in making the code easier to maintain, and we're going to
need the stat result for implementing `size()`, `mode()` etc.

Also move the classes from `pathlib._os` to `pathlib` proper.

gaogaotiantian committed:
gh-138860: Lazy import rlcompleter in pdb to avoid deadlock in subprocess (#139185)

corona10 committed:
gh-133171: Re-enable JUMP_BACKWARD to free-threading build (gh-137800)

StanFromIreland committed:
gh-83336: Add alias for consistency to `utf-8-sig` (#136530)

Closes #83336

picnixz committed:
gh-133644: remove `PyWeakref_GetObject` and `PyWeakref_GET_OBJECT` (GH-133657)

rokm committed:
gh-139231: Fix estimation of available stack size for recursion limit on macOS (GH-139232)

Use `pthread_get_stackaddr_np()` and `pthread_get_stacksize_np()` to determine the stack address and size.

sobolevn committed:
Remove `.. deprecated-removed` notes for `typing.{NamedTuple, TypedDict}` (#139298)

vstinner committed:
gh-139156: Use PyBytesWriter in _PyUnicode_EncodeCharmap() (#139251)

Replace PyBytes_FromStringAndSize() and _PyBytes_Resize() with the
PyBytesWriter API.

Add _PyBytesWriter_GetSize() and _PyBytesWriter_GetData() static
inline functions.

vstinner committed:
gh-139156: Use PyBytesWriter in PyUnicode_EncodeCodePage() (#139259)

Replace PyBytes_FromStringAndSize() and _PyBytes_Resize() with the
PyBytesWriter API.

richardhob committed:
Merge remote-tracking branch 'origin/gh42765-timit-execution-environment' into gh42765-timit-execution-environment
